pkg/operator/controller/controller.go

```diff
@@ -363,6 +295,203 @@ func (r *reconciler) ensureRouterDeleted(ci *ingressv1alpha1.ClusterIngress) err
 	return nil
 }
 
+// ensureRouterClusterRole ensures that the router cluster role exists and
+// matches the expected role.
+func (r *reconciler) ensureRouterClusterRole() error {
+	cr, err := r.ManifestFactory.RouterClusterRole()
+	if err != nil {
+		return fmt.Errorf("failed to build router cluster role: %v", err)
+	}
+
+	current := cr.DeepCopy()
+	key := types.NamespacedName{Namespace: cr.Namespace, Name: cr.Name}
+
+	if _, err := r.ensureRouterAsset(key, current); err != nil {
+		return err
+	}
+
+	// TODO: should we check for label/annotation changes or ignore them?
+	modified := !reflect.DeepEqual(current.Labels, cr.Labels) || !reflect.DeepEqual(current.Annotations, cr.Annotations)
+	modified = modified || !reflect.DeepEqual(current.Rules, cr.Rules)
+	modified = modified || !reflect.DeepEqual(current.AggregationRule, cr.AggregationRule)
+	if modified {
+		if err := r.Client.Update(context.TODO(), cr); err != nil {
+			return err
+		}
+
+		logrus.Infof("updated router cluster role %s", key.String())
+		return nil
+	}
+
+	return nil
+}
+
+// ensureRouterNamespaceAsset ensures that the router namespace exists and
+// matches the expected namespace.
+func (r *reconciler) ensureRouterNamespaceAsset() error {
+	ns, err := r.ManifestFactory.RouterNamespace()
+	if err != nil {
+		return fmt.Errorf("failed to build router namespace: %v", err)
+	}
+
+	current := ns.DeepCopy()
+	key := types.NamespacedName{Name: ns.Name}
+
+	if _, err := r.ensureRouterAsset(key, current); err != nil {
+		return err
+	}
+
+	// TODO: should we check for all label/annotation changes or just
+	//       specific ones node-selector/run-level?
+	modified := !reflect.DeepEqual(current.Labels, ns.Labels) || !reflect.DeepEqual(current.Annotations, ns.Annotations)
+	modified = modified || !reflect.DeepEqual(current.Spec, ns.Spec)
+	if modified {
+		if err := r.Client.Update(context.TODO(), ns); err != nil {
+			return err
+		}
+
+		logrus.Infof("updated router namespace %s", ns.Name)
+		return nil
+	}
+
+	return nil
+}
+
+// ensureRouterServiceAccount ensures that the router service account exists and
+// matches the expected service account.
+func (r *reconciler) ensureRouterServiceAccount() error {
+	sa, err := r.ManifestFactory.RouterServiceAccount()
+	if err != nil {
+		return fmt.Errorf("failed to build router service account: %v", err)
+	}
+
+	current := sa.DeepCopy()
+	key := types.NamespacedName{Namespace: sa.Namespace, Name: sa.Name}
+
+	if _, err := r.ensureRouterAsset(key, current); err != nil {
+		return err
+	}
+
+	// TODO: Should we check for Secrets, ImagePullSecrets and
+	//       AutomountServiceAccountToken - we don't really use those.
+	// TODO: should we check for label/annotation changes or ignore them?
+	modified := !reflect.DeepEqual(current.Labels, sa.Labels) || !reflect.DeepEqual(current.Annotations, sa.Annotations)
```

Comment:
This is an annoying one since the struct has no Spec to reflectively compare... is there some util in library-go or elsewhere for a generic "has anything mutable changed" comparisons?